Mathlib/Tactic/Ring.lean

```diff
@@ -0,0 +1,490 @@
+/-
+Copyright (c) 2018 Mario Carneiro. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Mario Carneiro
+-/
+
+import Lean.Elab.Tactic.Basic
+import Mathlib.Algebra.Ring.Basic
+import Mathlib.Tactic.NormNum
+
+/-!
+# `ring`
+
+Evaluate expressions in the language of commutative (semi)rings.
+Based on <http://www.cs.ru.nl/~freek/courses/tt-2014/read/10.1.1.61.3041.pdf> .
+-/
+
+open Lean Parser.Tactic Elab Command Elab.Tactic Meta
+
+open Expr in
+private def getAppFnAndArgsAux : Expr → Array Expr → Nat → Option (Name × Array Expr)
+  | app f a _,   as, i => getAppFnAndArgsAux f (as.set! i a) (i-1)
+  | const n _ _, as, i => some (n,as)
+  | _,           as, _ => none
+
+def Lean.Expr.getAppFnAndArgs (e : Expr) : Option (Name × Array Expr) :=
+  let dummy := mkSort levelZero
+  let nargs := e.getAppNumArgs
+  getAppFnAndArgsAux e (mkArray nargs dummy) (nargs-1)
```

Comment:
```suggestion
  e.withApp fun (fn, as) =>
    match fn with
    | const n _ _ => (n, as)
    | _ => none
```

Reply:
Thank you I will just delete this whole function and use `withApp` which I did not know